### PR TITLE
Permit transfers of wrapped ERC20 tokens between different addresses

### PR DESCRIPTION
**Pull request by james-chf**
Closes https://github.com/anoma/namada/issues/427

Currently, the EthBridge VP rejects any transactions which attempt to change keys in its space. This PR will change it to allow valid transfers of wrapped ERC20s between different users.

This PR does not include authorization of transfers by the user whose balance is decreasing - this will come in a later PR (https://github.com/anoma/namada/issues/432). End-to-end tests for CI will also come in a separate PR (it will be based on the shell example in https://github.com/anoma/namada/pull/394).

### CODE REVIEW

**Comment by james-chf** [comment]
pls update wasm

**Comment by james-chf** [comment]
pls update wasm

File: shared/src/ledger/eth_bridge/vp/mod.rs
Comment (batconjurer) [inline shared/src/ledger/eth_bridge/vp/mod.rs:70]:
```suggestion
            "Ethereum Bridge VP triggered",
```

File: shared/src/ledger/eth_bridge/vp/mod.rs
Comment (batconjurer) [inline shared/src/ledger/eth_bridge/vp/mod.rs:202]:
```suggestion
    let balance_a_pre = match reader.read_pre::<Amount>(&balance_a)?.unwrap_or_default()
```

File: shared/src/ledger/eth_bridge/vp/mod.rs
Comment (batconjurer) [inline shared/src/ledger/eth_bridge/vp/mod.rs:219]:
```suggestion
    let balance_b_pre = match reader.read_pre::<Amount>(&balance_b)? 
    .unwrap_or_default()
    .change();
```

File: shared/src/ledger/eth_bridge/vp/mod.rs
Comment (batconjurer) [inline shared/src/ledger/eth_bridge/vp/mod.rs:156]: The name of this method is not very accurate. This mainly checks that balances were adjust correctly,

File: shared/src/ledger/eth_bridge/vp/authorize.rs
Comment (batconjurer) [inline shared/src/ledger/eth_bridge/vp/authorize.rs:8]: I'm trying to follow the idea here. Is that idea that since the users vps won't be triggered otherwise, we need to manually trigger them to see if they approve a multitoken transfer? If so, shouldn't we pass in the verifiers set here and insert the necessary vps?

File: shared/src/ledger/eth_bridge/vp/store.rs
Comment (batconjurer) [inline shared/src/ledger/eth_bridge/vp/store.rs:34]:
```suggestion
       maybe_bytes.map(| ref bytes | 
           T::try_from_slice(bytes).wrap_err_with(|| "couldn't deserialize".to_string())
       )
       .transpose()
```

File: shared/src/ledger/eth_bridge/vp/store.rs
Comment (batconjurer) [inline shared/src/ledger/eth_bridge/vp/store.rs:13]: Since I've added these `read_pre_value` and `read_post_value` to the `Ctx` type in the bridge pool PR, let's just use those instead (just steal them). They return a simpler type (`Option<T>` instead of `Result<Option<T>>`) so it is simpler to handle.

**Review by batconjurer — CHANGES_REQUESTED** [review]

File: shared/src/ledger/eth_bridge/vp/mod.rs
Comment (james-chf) [reply shared/src/ledger/eth_bridge/vp/mod.rs]: Yeah this was not well named, have updated the name + docstring

File: shared/src/ledger/eth_bridge/vp/store.rs
Comment (james-chf) [reply shared/src/ledger/eth_bridge/vp/store.rs:13]: We should share code functionality, though I'd rather we make an issue to update the `Ctx` API in `main` and then refactor everywhere at once. We shouldn't silently ignore errors though if they happen (i.e. we should use `Result<Option<T>>` rather than `Option<T>`).

**Comment by james-chf** [comment]
pls update

**Comment by james-chf** [comment]
pls update wasm

File: shared/src/ledger/eth_bridge/vp/store.rs
Comment (batconjurer) [reply shared/src/ledger/eth_bridge/vp/store.rs:13]: These methods are in `eth_bridge_integration` now, so lets use them. But I agree, we shouldn't siliently ignore errors. It should be super fast to refactor the ctx methods to return the `Result<Option<T>>`

File: shared/src/ledger/eth_bridge/vp/authorize.rs
Comment (james-chf) [reply shared/src/ledger/eth_bridge/vp/authorize.rs:8]: > Implicit VPs are required for Namada launch, so I think triggering the necessary sender and receiver vps manually in this vp is the way to go.

Implicit VP issue - https://github.com/anoma/namada/issues/8

I will scope out https://github.com/anoma/namada/issues/432 further. There are some issues I think with the implicit VP approach in general, not directly related to the eth bridge, I can write up. For now then we can at least get parity with what works for normal tokens (i.e. proper authorization for transfers sent from an established address). Authorization for transfers sent from implicit addresses does not currently work even for normal token transfers using `namadac transfer`, we will need to wait for https://github.com/anoma/namada/issues/8 to be implemented

File: shared/src/ledger/eth_bridge/vp/store.rs
Comment (james-chf) [reply shared/src/ledger/eth_bridge/vp/store.rs:13]: sure I can refactor there in our codebase quickly, I will make an issue before merging this PR to update the `Ctx` API in `main`. I need to look closer at why it returns `Result<Option<T>>` and not just `Result<T>` as at a glance it looks like these `read_pre`/`read_post` methods error if a key isn't present in storage, I would have thought in that case it would have defaulted to `None` and the errors would be reserved for things like e.g. rocksdb or OS-level errors.

File: shared/src/ledger/eth_bridge/bridge_pool_vp.rs
Comment (james-chf) [inline shared/src/ledger/eth_bridge/bridge_pool_vp.rs:72]: This behaves the same way as before except it will log the error that happened in addition to returning `None`

Comment (batconjurer) [reply shared/src/ledger/eth_bridge/bridge_pool_vp.rs:72]: Looks good!

**Review by batconjurer — APPROVED** [review]

File: shared/src/ledger/eth_bridge/bridge_pool_vp.rs
Comment (sug0) [inline shared/src/ledger/eth_bridge/bridge_pool_vp.rs:130]: hm why do we need the explicit borrow here, on `ctx`? was the compiler (or clippy) complaining about something?

File: shared/src/ledger/eth_bridge/storage/mod.rs
Comment (sug0) [inline shared/src/ledger/eth_bridge/storage/mod.rs:52]: good candidates for a proptest :D

**Review by sug0 — COMMENTED** [review]
lgtm. just some random comments

**Review by sug0 — APPROVED** [review]

File: shared/src/ledger/eth_bridge/storage/mod.rs
Comment (james-chf) [reply shared/src/ledger/eth_bridge/storage/mod.rs:52]: will have a go at this (though not in this PR)

File: shared/src/ledger/eth_bridge/bridge_pool_vp.rs
Comment (james-chf) [reply shared/src/ledger/eth_bridge/bridge_pool_vp.rs:130]: It's because the trait this method comes from is implemented only for `&Ctx` and not `Ctx`

https://github.com/anoma/namada/blob/635f68cbf39656c2390d784f28104a98e57ed991/shared/src/ledger/native_vp.rs#L337

If there is a nice way to implement for both `Ctx` and `&Ctx` without too much duplication, I'd like to add it (though not in this PR) for ergonomics. (This `StorageReader` thing actually needs to be contributed back to `main` or may even be able to be replaced by the new native storage traits coming in soon)

**Comment by james-chf** [comment]
pls update wasm

**Pull request merged**
